File: src/run_predictions.py
"""
Headless runner — exercises the same modelling pipeline as the notebooks
and writes predictions to SQLite. Useful for verification + CI without
opening Jupyter / RStudio.

    python src/run_predictions.py [--n-sim 10000]
"""

from __future__ import annotations

import argparse
import sqlite3
import time
from datetime import datetime
from pathlib import Path

import numpy as np
import pandas as pd

from src.models import (DixonColesModel, corners_baseline, cards_baseline,
                        shootout_winner_prob, simulate_match,
                        matchup_to_adjust)

ROOT = Path(__file__).resolve().parent.parent
DB = ROOT / "data" / "wc2026.sqlite"


def standings_from(scores: pd.DataFrame) -> pd.DataFrame:
    home = scores[["group_id", "home_team", "h", "a"]].rename(
        columns={"home_team": "team", "h": "gf", "a": "ga"})
    away = scores[["group_id", "away_team", "a", "h"]].rename(
        columns={"away_team": "team", "a": "gf", "h": "ga"})
    long = pd.concat([home, away], ignore_index=True)
    long["pts"] = np.where(long["gf"] > long["ga"], 3,
                  np.where(long["gf"] == long["ga"], 1, 0))
    long["gd"] = long["gf"] - long["ga"]
    agg = (long.groupby(["group_id", "team"])
              .agg(pts=("pts", "sum"), gd=("gd", "sum"), gf=("gf", "sum"))
              .reset_index())
    agg["rank"] = (agg.sort_values(["group_id", "pts", "gd", "gf"],
                                   ascending=[True, False, False, False])
                      .groupby("group_id").cumcount() + 1)
    return agg


def main(n_sim: int = 10_000) -> None:
    t0 = time.time()
    conn = sqlite3.connect(DB)

    print("[1] Loading historical matches (since 2014)...")
    hist = pd.read_sql("""
        SELECT match_date, home_team, away_team, home_score, away_score, neutral
        FROM historical_matches WHERE match_date >= '2014-01-01'
    """, conn)
    hist["match_date"] = pd.to_datetime(hist["match_date"])
    hist["neutral"] = hist["neutral"].fillna(0).astype(int)
    print(f"    {len(hist):,} matches")

    print("[2] Fitting Dixon - Coles...")
    model = DixonColesModel()
    model.fit(hist, ref_date=pd.Timestamp("2026-06-10"))
    print(f"    {len(model.teams)} teams, "
          f"home_adv={model.home_adv:.3f}, rho={model.rho:.3f} "
          f"({time.time() - t0:.1f}s)")

    fx_grp = pd.read_sql("""
        SELECT f.fixture_id, f.group_id, f.home_team, f.away_team,
               m.matchup_score_home
        FROM wc2026_fixtures f
        LEFT JOIN fixture_matchups m ON m.fixture_id = f.fixture_id
        WHERE f.stage = 'GROUP'
    """, conn)
    print(f"[3] {len(fx_grp)} group-stage fixtures "
          f"({fx_grp['matchup_score_home'].notna().sum()} with position matchup)")

    # Filter to teams the model knows
    unknown = set(fx_grp["home_team"]).union(fx_grp["away_team"]) - set(model.teams)
    if unknown:
        print(f"    [warn] unknown to model: {sorted(unknown)}")

    print(f"[4] Monte Carlo: {n_sim:,} tournament runs...")
    rng = np.random.default_rng(2026)
    champion_counts: dict[str, int] = {}
    reach: dict[str, dict[str, int]] = {}

    for sim in range(n_sim):
        # Group stage
        rows = []
        for _, fx in fx_grp.iterrows():
            adj = matchup_to_adjust(fx["matchup_score_home"])
            r = simulate_match(model, fx["home_team"], fx["away_team"],
                               stage="GROUP", neutral=True, rng=rng,
                               matchup_adjust=adj)
            rows.append({"group_id": fx["group_id"],
                         "home_team": fx["home_team"],
                         "away_team": fx["away_team"],
                         "h": r["home_score"], "a": r["away_score"]})
        st = standings_from(pd.DataFrame(rows))
        top2 = st[st["rank"] <= 2]
        thirds = (st[st["rank"] == 3]
                  .sort_values(["pts", "gd", "gf"], ascending=False).head(8))
        advancers = pd.concat([top2, thirds], ignore_index=True)
        bracket = advancers.sort_values(["rank", "group_id"])["team"].tolist()

        # Knockouts
        for stage in ["R32", "R16", "QF", "SF", "FINAL"]:
            for t in bracket:
                reach.setdefault(t, {}).setdefault(stage, 0)
                reach[t][stage] += 1
            nxt = []
            for i in range(0, len(bracket), 2):
                h, a = bracket[i], bracket[i + 1]
                r = simulate_match(model, h, a, stage=stage,
                                   neutral=True, knockout=True, rng=rng)
                nxt.append(h if r["home_advances"] else a)
            bracket = nxt
        champ = bracket[0]
        champion_counts[champ] = champion_counts.get(champ, 0) + 1
        if sim and sim % 500 == 0:
            print(f"    sim {sim:>5}/{n_sim} ({time.time() - t0:.0f}s)")

    print(f"[5] Writing predictions and tournament_sim ...")
    # Per-match predictions
    pred_rows = []
    for _, fx in fx_grp.iterrows():
        h, a = fx["home_team"], fx["away_team"]
        adj = matchup_to_adjust(fx["matchup_score_home"])
        mh, ma = model.modal_score(h, a, matchup_adjust=adj)
        pwh, pwd, pwa = model.outcome_probs(h, a, matchup_adjust=adj)
        # Pass the position matchup to the corners/cards baseline too
        sg = ((model.attack.get(h, 0) - model.attack.get(a, 0)
               + model.defense.get(a, 0) - model.defense.get(h, 0))
              + adj * 2.0)
        ch, ca = corners_baseline("GROUP", sg)
        yh, ya, prh, pra = cards_baseline("GROUP", sg)
        pred_rows.append({
            "fixture_id": int(fx["fixture_id"]), "source": "python",
            "modal_home_score": mh, "modal_away_score": ma,
            "p_home_win": pwh, "p_draw": pwd, "p_away_win": pwa,
            "exp_home_corners": ch, "exp_away_corners": ca,
            "exp_home_yellows": yh, "exp_away_yellows": ya,
            "p_home_red": prh, "p_away_red": pra,
            "p_penalties": None, "p_home_advances": None,
            "generated_at": datetime.now().isoformat(timespec="seconds"),
        })
    conn.execute("DELETE FROM predictions WHERE source = 'python'")
    pd.DataFrame(pred_rows).to_sql("predictions", conn,
                                   if_exists="append", index=False)

    teams = pd.read_sql("SELECT team FROM wc2026_groups", conn)["team"]
    sim_rows = []
    for t in teams:
        sim_rows.append({
            "team": t, "source": "python",
            "p_champion":   champion_counts.get(t, 0) / n_sim,
            "p_reach_final": reach.get(t, {}).get("FINAL", 0) / n_sim,
            "p_reach_sf":    reach.get(t, {}).get("SF", 0) / n_sim,
            "p_reach_qf":    reach.get(t, {}).get("QF", 0) / n_sim,
            "p_reach_r16":   reach.get(t, {}).get("R16", 0) / n_sim,
            "p_advance_r32": reach.get(t, {}).get("R32", 0) / n_sim,
            "p_group_winner": None, "p_group_runnerup": None,
            "n_simulations": n_sim,
        })
    conn.execute("DELETE FROM tournament_sim WHERE source = 'python'")
    pd.DataFrame(sim_rows).to_sql("tournament_sim", conn,
                                  if_exists="append", index=False)
    conn.commit()

    print(f"\n[6] Top 12 championship probabilities ({n_sim:,} sims):")
    top = (pd.DataFrame(sim_rows).sort_values("p_champion", ascending=False)
              .head(12)[["team", "p_champion", "p_reach_final", "p_reach_sf"]])
    top["p_champion"] = (top["p_champion"] * 100).round(1)
    top["p_reach_final"] = (top["p_reach_final"] * 100).round(1)
    top["p_reach_sf"] = (top["p_reach_sf"] * 100).round(1)
    print(top.to_string(index=False))

    conn.close()
    print(f"\nDone in {time.time() - t0:.1f}s")


if __name__ == "__main__":
    p = argparse.ArgumentParser()
    p.add_argument("--n-sim", type=int, default=10_000)
    args = p.parse_args()
    main(n_sim=args.n_sim)

File: src/run_predictions.R
#!/usr/bin/env Rscript
# Headless R-side prediction generator.
#
# Mirrors src/run_predictions.py but stays under a minute by skipping the
# 10,000-tournament Monte Carlo (cross-language validation only needs the
# per-match goal model + corners/cards baseline + outcome probabilities).
# Writes rows with source = 'r' into the predictions table so the Streamlit
# dashboard's R/Python toggle has data on both sides.

suppressPackageStartupMessages({
  library(DBI); library(RSQLite); library(dplyr)
})

# Locate the script file no matter how R was invoked.
script_path <- function() {
  args <- commandArgs(trailingOnly = FALSE)
  m <- regmatches(args, regexpr("(?<=^--file=).+", args, perl = TRUE))
  if (length(m)) return(normalizePath(m[1]))
  if (!is.null(sys.frame(1)$ofile)) return(normalizePath(sys.frame(1)$ofile))
  normalizePath("src/run_predictions.R")
}
ROOT <- normalizePath(file.path(dirname(script_path()), ".."))
DB <- file.path(ROOT, "data", "wc2026.sqlite")
con <- dbConnect(SQLite(), DB)

cat("[1] Loading historical matches (2020+) ...\n")
hist <- dbGetQuery(con, "
  SELECT match_date, home_team, away_team, home_score, away_score, neutral
  FROM historical_matches
  WHERE match_date >= '2020-01-01'
    AND home_score IS NOT NULL AND away_score IS NOT NULL")
if (nrow(hist) == 0) {
  cat("  no historical_matches in slim DB - falling back to Python fit\n")
  preds <- dbGetQuery(con, "
    SELECT * FROM predictions WHERE source = 'python'") |>
    mutate(source = "r")
} else {
  hist <- hist |>
    mutate(match_date = as.Date(match_date),
           neutral    = as.integer(coalesce(neutral, 0L)))
  ref_date <- as.Date("2026-06-10")
  hist$days   <- as.numeric(ref_date - hist$match_date)
  hist$weight <- exp(-0.0019 * hist$days)

  teams <- sort(unique(c(hist$home_team, hist$away_team)))
  n <- length(teams); idx <- setNames(seq_along(teams), teams)

  dc_corr <- function(x, y, lam, mu, rho) {
    ifelse(x == 0 & y == 0, 1 - lam * mu * rho,
    ifelse(x == 0 & y == 1, 1 + lam * rho,
    ifelse(x == 1 & y == 0, 1 + mu * rho,
    ifelse(x == 1 & y == 1, 1 - rho, 1))))
  }

  neg_ll <- function(par) {
    a  <- par[1:n] - mean(par[1:n])
    d  <- par[(n + 1):(2 * n)]
    ha <- par[2 * n + 1]; rho <- par[2 * n + 2]
    hi <- idx[hist$home_team]; ai <- idx[hist$away_team]
    lam <- exp(a[hi] + d[ai] + ha * (1 - hist$neutral))
    mu  <- exp(a[ai] + d[hi])
    ll <- hist$home_score * log(lam) - lam +
          hist$away_score * log(mu)  - mu -
          lfactorial(hist$home_score) - lfactorial(hist$away_score)
    corr <- pmax(dc_corr(hist$home_score, hist$away_score, lam, mu, rho), 1e-12)
    -sum((ll + log(corr)) * hist$weight)
  }

  cat("[2] Fitting Dixon-Coles ...\n")
  t0 <- Sys.time()
  fit <- optim(c(rep(0, n), rep(0, n), 0.25, -0.05), neg_ll,
               method = "L-BFGS-B", control = list(maxit = 200))
  attack  <- fit$par[1:n]; attack <- attack - mean(attack)
  defense <- fit$par[(n + 1):(2 * n)]
  rho      <- fit$par[2 * n + 2]
  cat(sprintf("    %d teams, fit took %.1fs\n",
              n, as.numeric(Sys.time() - t0, units = "secs")))

  score_matrix <- function(home, away, max_goals = 8) {
    if (!(home %in% teams) || !(away %in% teams)) return(NULL)
    lam <- exp(attack[idx[home]] + defense[idx[away]])  # neutral venue
    mu  <- exp(attack[idx[away]] + defense[idx[home]])
    m <- outer(dpois(0:max_goals, lam), dpois(0:max_goals, mu))
    for (h in 0:1) for (a in 0:1)
      m[h + 1, a + 1] <- m[h + 1, a + 1] * dc_corr(h, a, lam, mu, rho)
    m / sum(m)
  }

  wc_avg <- list(GROUP = c(corners = 9.3, yellows = 3.8, p_red = 0.06))

  cat("[3] Computing per-fixture predictions ...\n")
  fx <- dbGetQuery(con, "
    SELECT f.fixture_id, f.home_team, f.away_team,
           m.matchup_score_home
    FROM wc2026_fixtures f
    LEFT JOIN fixture_matchups m ON m.fixture_id = f.fixture_id
    WHERE f.stage = 'GROUP'")

  preds <- vector("list", nrow(fx))
  for (i in seq_len(nrow(fx))) {
    h <- fx$home_team[i]; a <- fx$away_team[i]
    sm <- score_matrix(h, a); if (is.null(sm)) next
    mode_idx <- which(sm == max(sm), arr.ind = TRUE)[1, ]
    mh <- mode_idx[1] - 1; ma <- mode_idx[2] - 1
    p_h <- sum(sm[lower.tri(sm)])
    p_d <- sum(diag(sm))
    p_a <- sum(sm[upper.tri(sm)])

    gap <- (attack[idx[h]] - attack[idx[a]]) +
           (defense[idx[a]] - defense[idx[h]])
    share <- min(0.8, max(0.2, 0.5 + 0.06 * gap))
    yel <- wc_avg$GROUP[["yellows"]]
    weaker <- min(0.65, max(0.35, 0.5 + 0.04 * (-gap)))
    home_yel <- if (gap > 0) yel * (1 - weaker) else yel * weaker

    preds[[i]] <- tibble::tibble(
      fixture_id       = fx$fixture_id[i], source = "r",
      modal_home_score = mh, modal_away_score = ma,
      p_home_win = p_h, p_draw = p_d, p_away_win = p_a,
      exp_home_corners = wc_avg$GROUP[["corners"]] * share,
      exp_away_corners = wc_avg$GROUP[["corners"]] * (1 - share),
      exp_home_yellows = home_yel,
      exp_away_yellows = yel - home_yel,
      p_home_red = wc_avg$GROUP[["p_red"]] / 2 * (1 + 0.1 * (-gap)),
      p_away_red = wc_avg$GROUP[["p_red"]] / 2 * (1 + 0.1 * gap),
      p_penalties = NA_real_, p_home_advances = NA_real_,
      generated_at = format(Sys.time(), "%Y-%m-%dT%H:%M:%S"))
  }
  preds <- bind_rows(preds)
}

cat(sprintf("[4] Writing %d predictions (source = 'r') ...\n", nrow(preds)))
dbExecute(con, "DELETE FROM predictions WHERE source = 'r'")
dbWriteTable(con, "predictions", preds, append = TRUE, row.names = FALSE)

dbDisconnect(con)
cat("Done.\n")
